pkg/constraints/constraint.go

```diff
@@ -0,0 +1,83 @@
+package constraints
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+)
+
+// OLMConstraintType is the schema "type" key for all constraints known to OLM
+// (except for legacy types).
+const OLMConstraintType = "olm.constraint"
+
+// Constraint holds parsed, potentially nested dependency constraints.
+type Constraint struct {
+	// Constraint message that surfaces in resolution
+	// This field is optional
+	Message string `json:"message"`
+
+	// The cel struct that contraints CEL expression
+	Cel *Cel `json:"cel,omitempty" yaml:"cel,omitempty"`
+
+	// Package defines a constraint for a package within a version range.
+	Package *PackageConstraint `json:"package,omitempty" yaml:"package,omitempty"`
+	// GVK defines a constraint for a GVK.
+	GVK *GVKConstraint `json:"gvk,omitempty" yaml:"gvk,omitempty"`
+
+	// All, Any, and None are compound constraints. See this enhancement for details:
+	// https://github.com/operator-framework/enhancements/blob/master/enhancements/compound-bundle-constraints.md
+	All *CompoundConstraint `json:"all,omitempty" yaml:"all,omitempty"`
+	Any *CompoundConstraint `json:"any,omitempty" yaml:"any,omitempty"`
+	// A note on None: this constraint is not particularly useful by itself.
+	// It should be used within an All constraint alongside some other constraint type
+	// since saying "none of these GVKs/packages/etc." without an alternative doesn't make sense.
+	None *CompoundConstraint `json:"none,omitempty" yaml:"none,omitempty"`
+}
+
+// CompoundConstraint holds a list of potentially nested constraints
+// over which a boolean operation is applied.
+type CompoundConstraint struct {
+	Constraints []Constraint `json:"constraints"`
+}
+
+// GVKConstraint defines a GVK constraint.
+type GVKConstraint struct {
+	Group   string `json:"group"`
+	Kind    string `json:"kind"`
+	Version string `json:"version"`
+}
+
+// PackageConstraint defines a package constraint.
+type PackageConstraint struct {
+	// Name of the package.
+	Name string `json:"name"`
+	// VersionRange required for the package.
+	VersionRange string `json:"versionRange"`
+}
+
+// maxConstraintSize defines the maximum raw size in bytes of an olm.constraint.
+// 64Kb seems reasonable, since this number allows for long description strings
+// and either few deep nestings or shallow nestings and long constraints lists,
+// but not both.
+const maxConstraintSize = 2 << 16
```

Comment:
Non-blocking: I am weary of hardcoded maximums that users cannot override.

Reply:
This is for security purposes so no one should be able to configure this, at least not for now. I'll add a note to come back to this later.